plutus-core/index-envs/src/Data/RandomAccessList/SkewBinary.hs

```diff
@@ -74,122 +76,94 @@ uncons = \case
-    indexTree treeSize offset (Node _ t1 t2) =
-        let halfSize = unsafeShiftR treeSize 1 -- probably faster than `div w 2`
-            offset' = offset - 1
-        in if offset' <= halfSize
-           then indexTree halfSize offset' t1
-           else indexTree halfSize (offset' - halfSize) t2
-
-{- Note [Optimizations of safeIndexOneCont]
-Bangs in the local definitions of 'safeIndexOneCont' are needed to tell GHC that the functions are
+{- Note [Optimizations of contIndexZero]
+Bangs in the local definitions of 'contIndexZero' are needed to tell GHC that the functions are
 strict in the 'Word64' argument, so that GHC produces workers operating on @Word64#@.
 
 The function itself is CPS-ed, so that the arguments force the local definitions to be retained
-within 'safeIndexOneCont' instead of being pulled out via full-laziness or some other optimization
-pass. This ensures that when 'safeIndexOneCont' gets inlined, the local definitions appear directly
-in the GHC Core, allowing GHC to inline the arguments of 'safeIndexOneCont' and transform the whole
+within 'contIndexZero' instead of being pulled out via full-laziness or some other optimization
+pass. This ensures that when 'contIndexZero' gets inlined, the local definitions appear directly
+in the GHC Core, allowing GHC to inline the arguments of 'contIndexZero' and transform the whole
 thing into a beautiful recursive join point full of @Word64#@s, i.e. allocating very little if
 anything at all.
 -}
 
--- See Note [Optimizations of safeIndexOneCont].
-safeIndexOneCont :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
-safeIndexOneCont z f = findTree where
+-- See Note [Optimizations of contIndexZero].
+contIndexZero :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
+contIndexZero z f = findTree where
     findTree :: RAList a -> Word64 -> b
-    -- See Note [Optimizations of safeIndexOneCont].
+    -- See Note [Optimizations of contIndexZero].
     findTree Nil !_ = z
     findTree (BHead w t ts) i =
-        if i <= w
+        if i < w
         then indexTree w i t
         else findTree ts (i-w)
 
     indexTree :: Word64 -> Word64 -> Tree a -> b
-    -- See Note [Optimizations of safeIndexOneCont].
-    indexTree !w 1 t = case t of
+    -- See Note [Optimizations of contIndexZero].
+    indexTree !w 0 t = case t of
         Node x _ _ -> f x
         Leaf x     -> if w == 1 then f x else z
-    indexTree _ 0 _ = z -- "index zero"
     indexTree _ _ (Leaf _) = z
     indexTree treeSize offset (Node _ t1 t2) =
         let halfSize = unsafeShiftR treeSize 1 -- probably faster than `div w 2`
-            offset' = offset - 1
-        in if offset' <= halfSize
-           then indexTree halfSize offset' t1
-           else indexTree halfSize (offset' - halfSize) t2
-{-# INLINE safeIndexOneCont #-}
+        in if offset <= halfSize
+           then indexTree halfSize (offset - 1) t1
+           else indexTree halfSize (offset - 1 - halfSize) t2
```

Comment:
Out of curiosity: did you inline `offset'` here because it is such a trivial computation?

Reply:
It was inlined in the original `safeIndexZero` and not inlined in the original `safeIndexOne`, I think because in the latter `offset'` was used twice for every branch, while in the former only once. So I followed and picked the non-inlined version. It doesn't really matter.